crates/bevy_ecs/macros/src/states.rs

```diff
@@ -28,15 +26,46 @@ pub fn derive_states(input: TokenStream) -> TokenStream {
     trait_path.segments.push(format_ident!("schedule").into());
     trait_path.segments.push(format_ident!("States").into());
     let struct_name = &ast.ident;
-    let idents = enumeration.variants.iter().map(|v| &v.ident);
-    let len = idents.len();
+    let fieldless_idents = enumeration
+        .variants
+        .iter()
+        .filter(|v| v.fields.is_empty())
+        .map(|v| &v.ident);
+    let fieldful_idents: Vec<&Ident> = enumeration
+        .variants
+        .iter()
+        .filter(|v| !v.fields.is_empty())
+        .map(|v| &v.ident)
+        .collect();
+    // kagamine
+    let len = enumeration.variants.len();
+    let (variants_impl, iter_type) = if fieldful_idents.len() > 0 {
+        (
+            quote! {
+                let mut fields = vec![#(Self::#fieldless_idents,)*];
+                let fieldful = [#(<Self::#fieldful_idents as States>::variants().map(|variant| {
+                            Self::#fieldful_idents(variant)
+                }),)*];
+                for field in fieldful {
+                    fields.extend(field)
+                }
```

Comment:
I feel like you could probably use [`flat_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map) here and remove the extra array and for loop, right?

Reply:
Ah, okay

Reply:
I'm not sure if flat_map is the right option? I changed to `flatten` though.

Reply:
![A screenshot of Cargo lint complaining about not using flat_map](https://user-images.githubusercontent.com/66000635/227464988-04cf15fa-3f2b-4e16-9362-37c8f9284d49.png)
Ah, I see what you mean.

Reply:
Wait, this was different though.